GetIntoTeachingApi/Models/BaseModel.cs

```diff
@@ -260,8 +260,23 @@ private void MapRelationshipAttributesToEntity(Entity source, ICrmService crm, O
                 var attribute = EntityRelationshipAttribute(property);
                 var value = property.GetValue(this);
 
-                if (attribute == null || value == null)
+                if (attribute == null)
+                {
+                    continue;
+                }
+
+                var relationship = new Relationship(attribute.Name);
+
+                if (value == null)
                 {
+                    context.LoadProperty(source, relationship);
+
+                    if (source.RelatedEntities.ContainsKey(relationship))
+                    {
+                        Entity relatedEntity = source.RelatedEntities.First(entity => entity.Key == relationship).Value.Entities.First();

```

Comment:
This would only work when there can only be one related entity (e.g. event -> building) 😑

Reply:
So looking into this a bit the issue with doing it in the `BaseModel` is that we're not able to distinguish between all the possible scenarios. If you take, for example, a candidate record being saved with a single qualification that could either be interpreted as:

- Just add the single qualification
- Add the single qualification and remove all the other qualifications (as they're not present in our version of the relationship)

The only way to clear up the ambiguity would be to ensure that when you save a candidate the current state of the other qualifications is also present and up to date.

We would need to ensure this further up the chain, before the model mapping logic is called - probably as part of the `UpsertCandidateJob#SaveCandidate` - so load in the existing qualifications/past teaching positions here. We would then also need to query the current state of qualifications in the mapping logic so that we can do a diff and unlink any that are no longer present.

We would also need to check the change tracking as we do when we map an individual field attribute so that we can differentiate between `null` (all removed) and `not changed`.

I think this would be overcomplicated and difficult to maintain. A simpler approach may be to hook into the mapping process from the `TeachingEvent` class and perform a check/unlink the relationship there when its being mapped:

```
public bool ShouldMap(ICrmService crm, Context context) {
  if (buildingRemoved)
  {
    crm.RemoveLink(..., context)
  }

  return true;
}
```